Slim/Control/Queries.pm

```diff
@@ -1989,6 +1995,147 @@ sub irenableQuery {
 	$request->setStatusDone();
 }
 
+sub labelsQuery {
+	my $request = shift;
+
+	# check this is the correct query.
+	if ($request->isNotQuery([['labels']])) {
+		$request->setStatusBadDispatch();
+		return;
+	}
+
+	if (!Slim::Schema::hasLibrary()) {
+		$request->setStatusNotDispatchable();
+		return;
+	}
+
+	my $sqllog = main::DEBUGLOG && logger('database.sql');
+
+	# get our parameters
+	my $client        = $request->client();
+	my $index         = $request->getParam('_index');
+	my $quantity      = $request->getParam('_quantity');
+	my $year          = $request->getParam('year');
+	my $contributorID = $request->getParam('artist_id');
+	my $workID        = $request->getParam('work_id');
+	my $libraryID     = Slim::Music::VirtualLibraries->getRealId($request->getParam('library_id'));
+	my $tags          = $request->getParam('tags') || '';
+
+	my $sql  = 'SELECT distinct label FROM albums ';
+	my $w    = [];
+	my $p    = [];
+	push @{$w}, 'albums.label IS NOT NULL';
+
+	if (defined $contributorID) {
+
+	$sql .= 'JOIN contributor_album ON albums.id = contributor_album.contributor ';
+		# handle the case where we're asked for the VA id => return compilations
+		if ($contributorID == Slim::Schema->variousArtistsObject->id) {
+			push @{$w}, 'albums.compilation = ?';
+			push @{$p}, 1;
+		}
+		else {
+			push @{$w}, 'contributor_album.contributor = ?';
+			push @{$p}, $contributorID;
+		}
+	}
+
+	if ( $libraryID ) {
+		$sql .= 'JOIN library_album ON library_album.album = albums.id ';
+		push @{$w}, 'library_album.library = ?';
+		push @{$p}, $libraryID;
+	}
+
+	if (defined $year) {
+		push @{$w}, 'albums.year = ?';
+		push @{$p}, $year;
+	}
+	if (defined $workID) {
+		$sql .= "JOIN tracks on tracks.album = albums.id ";
+		if ( $workID eq "-1" ) {
+			push @{$w}, 'tracks.work IS NOT NULL';
+		} else {
+			push @{$w}, 'tracks.work = ?';
+			push @{$p}, $workID;
+		}
+	}
+
+	if ( @{$w} ) {
+		$sql .= 'WHERE ';
+		my $s = join( ' AND ', @{$w} );
+		$s =~ s/\%/\%\%/g;
+		$sql .= $s . ' ';
+	}
+
+	my $dbh = Slim::Schema->dbh;
+
+	my $stillScanning = Slim::Music::Import->stillScanning();
+
+	# Get count of all results, the count is cached until the next rescan done event
+	my $cacheKey = md5_hex($sql . join( '', @{$p} ) . Slim::Music::VirtualLibraries->getLibraryIdForClient($client));
+
+	my $count = $cache->{$cacheKey};
+	if ( !$count ) {
+		my $total_sth = $dbh->prepare_cached( qq{
+			SELECT COUNT(1) FROM ( $sql ) AS t1
+		} );
+
+		$total_sth->execute( @{$p} );
+		($count) = $total_sth->fetchrow_array();
+		$total_sth->finish;
+
+		if ( !$stillScanning ) {
+			$cache->{$cacheKey} = $count;
+		}
+	}
+
+	# now build the result
+
+	if ($stillScanning) {
+		$request->addResult('rescan', 1);
+	}
+
+	$count += 0;
+
+	my ($valid, $start, $end) = $request->normalize(scalar($index), scalar($quantity), $count);
+
+	if ($valid && $tags ne 'CC') {
+
+		$sql .= 'ORDER BY 1 ';
```

Comment:
Wow... hadn't known this shortcut. But... please use the column name for easier reading.